UnsupportedOperationException in RealmBaseAdapter

I get such exception after I guess last change regarding notifications (0.86.0-SNAPSHOT):

`Caused by: java.lang.UnsupportedOperationException`
`		at java.util.concurrent.CopyOnWriteArrayList$CowIterator.remove(CopyOnWriteArrayList.java:744)`
`		at io.realm.HandlerController.addChangeListenerAsWeakReference(HandlerController.java:145)`
`		at io.realm.RealmBaseAdapter.<init>(RealmBaseAdapter.java:56)`
`		at com.borg.forza.adapters.RealmModelAdapter.<init>(RealmModelAdapter.java:0)`

Seems like CopyOnWriteArrayList iterator does not support remove operation.

--- a/realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java
@@ -495,6 +495,45 @@ public class NotificationsTest extends AndroidTestCase {
         assertEquals(0, realm.handlerController.weakChangeListeners.size());
     }
 
+
+    // Test that that a WeakReferenceListener can be removed.
+    // This test is not a proper GC test, but just ensures that listeners can be removed from the list of weak listeners
+    // without throwing an exception.
+    public void testRemovingWeakReferenceListener() throws InterruptedException {
+        final AtomicInteger counter = new AtomicInteger(0);
+        realm = Realm.getInstance(getContext());
+        RealmChangeListener listenerA = new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                counter.incrementAndGet();
+            }
+        };
+        RealmChangeListener listenerB = new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                counter.incrementAndGet();
+            }
+        };
+        realm.handlerController.addChangeListenerAsWeakReference(listenerA);
+
+        // There is no guaranteed way to release the WeakReference,
+        // just clear it.
+        for (WeakReference<RealmChangeListener> weakRef : realm.handlerController.weakChangeListeners) {
+            weakRef.clear();
+        }
+
+        realm.handlerController.addChangeListenerAsWeakReference(listenerB);
+
+        realm.beginTransaction();
+        realm.createObject(AllTypes.class);
+        realm.commitTransaction();
+
+        assertEquals(1, counter.get());
+        assertEquals(1, realm.handlerController.weakChangeListeners.size());
+    }
+
+
+
     // Tests that if the same configuration is used on 2 different Looper threads that each gets its own Handler. This
     // prevents commitTransaction from accidentally posting messages to Handlers which might reference a closed Realm.
     public void testDoNotUseClosedHandler() throws InterruptedException {

--- a/realm/realm-library/src/main/java/io/realm/HandlerController.java
+++ b/realm/realm-library/src/main/java/io/realm/HandlerController.java
@@ -139,17 +139,32 @@ public class HandlerController implements Handler.Callback {
      */
     void addChangeListenerAsWeakReference(RealmChangeListener listener) {
         Iterator<WeakReference<RealmChangeListener>> iterator = weakChangeListeners.iterator();
+        List<WeakReference<RealmChangeListener>> toRemoveList = null;
+        boolean addListener = true;
         while (iterator.hasNext()) {
-            WeakReference<RealmChangeListener> ref = iterator.next();
-            if (ref.get() == null) {
-                iterator.remove();
+            WeakReference<RealmChangeListener> weakRef = iterator.next();
+            RealmChangeListener weakListener = weakRef.get();
 
-            } else if (ref.get() == listener) {
-                return; // It has already been added before
+            // Collect all listeners that are GC'ed
+            if (weakListener == null) {
+                if (toRemoveList == null) {
+                    toRemoveList = new ArrayList<WeakReference<RealmChangeListener>>(weakChangeListeners.size());
+                }
+                toRemoveList.add(weakRef);
+            }
+
+            // Check if Listener already exists
+            if (weakListener == listener) {
+                addListener = false;
+                break;
             }
         }
-
-        weakChangeListeners.add(new WeakReference<RealmChangeListener>(listener));
+        if (toRemoveList != null) {
+            weakChangeListeners.removeAll(toRemoveList);
+        }
+        if (addListener) {
+            weakChangeListeners.add(new WeakReference<RealmChangeListener>(listener));
+        }
     }
 
     void removeChangeListener(RealmChangeListener listener) {

--- a/realm/realm-library/src/main/java/io/realm/HandlerController.java
+++ b/realm/realm-library/src/main/java/io/realm/HandlerController.java
@@ -156,7 +156,6 @@ public class HandlerController implements Handler.Callback {
             // Check if Listener already exists
             if (weakListener == listener) {
                 addListener = false;
-                break;
             }
         }
         if (toRemoveList != null) {

